README.md
# Money Type

[![TravisCI](https://travis-ci.org/phpexpertsinc/MoneyType.svg?branch=master)](https://travis-ci.org/phpexpertsinc/MoneyType)
[![Maintainability](https://api.codeclimate.com/v1/badges/37f9c3f6a89b5a72256e/maintainability)](https://codeclimate.com/github/phpexpertsinc/MoneyType/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/37f9c3f6a89b5a72256e/test_coverage)](https://codeclimate.com/github/phpexpertsinc/MoneyType/test_coverage)

MoneyType is a PHP Experts, Inc., Project meant to keep track of money precisely, even up to millions of dollars.

It is never safe to store money as floats or even integers times one hundred. This project gives you the security
you need, so that something like the bug in ["Office Space"](https://www.imdb.com/title/tt0151804/) just won't
happen to your enterprise application.

## Installation

Via Composer

```bash
composer require phpexpertsinc/money-type
```

## Usage

```php
use PHPExperts\MoneyType\Money;

$money = new Money(5.22);

$money->add(0.55);
echo "$money\n"; // 5.77

# It keeps precision much much better than mere cents.
$money->subtract(0.0001);
echo "$money\n"; // 5.77
$money->subtract(0.004);
echo "$money\n"; // 5.77
$money->subtract(0.001);
echo "$money\n"; // 5.76

# Actually, to the tenth decimal place.
# So it's Cryptocurrency-ready!
$money->multiply(55.7773);
echo "$money\n"; // 321.55

# But deep down, it stores the true values to many decimals (if you use BCMath).
echo $money->getWithFullPrecision() . "\n"; // 321.5508738395

$money->divide('1.000005');
echo "$money\n"; // 321.55
echo $money->getWithFullPrecision() . "\n"; // 321.5492660931

# You can also compare really large numbers with one another, up to 10 decimal places.
# PHP Experts' MoneyType is cryptocurrency ready! In fact, that's what it was designed for!
$money->compare(321.5492660931);       // 0 = equal
$money->compare(321.549266093009);     // -1 = less
$money->compare(321.5492660931000001); // 1 = more

# Get the object.
print_r($money);

# It is cryptocurrency ready:
# Converts Bitcoins to Satoshis
$btc = '1.55527331';
$satoshis = NumberHelper::convertToCents($btc, 8); // 155527331 (int)
```

# Use cases

PHPExperts\MoneyType\Money  
 ✔ Can only be instantiated with a numeric string  
 ✔ Will report what strategy is being used  
 ✔ Will use BCMath if it is available  
 ✔ Will fall back to native php if necessary  
 ✔ Proxies everything to its calculation strategy  
 ✔ Confirm that the readme demo works

PHPExperts\MoneyType\Internal\BCMathCalcStrategy  
 ✔ Can get the full precision value to more than sixteen decimals  
 ✔ Can add with high precision  
 ✔ Can subtract with high precision  
 ✔ Can multiply with high precision  
 ✔ Can divide with high precision  
 ✔ Can compare two numbers with high precision  
 ✔ Can compute high precision modulus  
 ✔ Can compute the modulus of decimals  
 ✔ Can round with high precision  
 ✔ Can only be instantiated with a numeric string  
 ✔ Access the object as a string to get its valuation  
 ✔ Can add with cent precision  
 ✔ Can subtract with cent precision  
 ✔ Can multiply with cent precision  
 ✔ Can divide with cent precision  
 ✔ Can compare two numbers with cent precision  
 ✔ Will not accept a non number for any operation

PHPExperts\MoneyType\Internal\NativeCalcStrategy  
 ✔ Wont attempt operations with non numbers  
 ✔ Can get the full precision value to two decimals  
 ✔ Cannot compute the modulus of decimals  
 ✔ Will throw an exception if asked to compute an integer modulus  
 ✔ Can compute the modulus of integers if dev passes i am a dummy parameter  
 ✔ Can only be instantiated with a numeric string  
 ✔ Access the object as a string to get its valuation  
 ✔ Can add with cent precision  
 ✔ Can subtract with cent precision  
 ✔ Can multiply with cent precision  
 ✔ Can divide with cent precision  
 ✔ Can compare two numbers with cent precision  
 ✔ Will not accept a non number for any operation

PHPExperts\MoneyType\Internal\NumberHelper: A collection of functions for number manipulation.  
 ✔ Will return true if given a float  
 ✔ Will return true if given a float string  
 ✔ Will return false if given an integer  
 ✔ Will return false if given an integer string  
 ✔ Will throw an exception if given anything else  
 ✔ Can convert dollars to cents integer without precision loss  
 ✔ Can convert bitcoins to satoshis

## Testing

```bash
phpunit
```

# Contributors

[Theodore R. Smith](https://www.phpexperts.pro/]) <theodore@phpexperts.pro>  
GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690  
CEO: PHP Experts, Inc.

[[Alix Axel](https://stackoverflow.com/users/89771/alix-axel)]  
who contributed the base version of our "bcround()" function
from https://stackoverflow.com/a/1653826/430062.

## License

MIT license. Please see the [license file](LICENSE) for more information.



CHANGELOG.md
## v2.1.0: 2023-07-23

* **[2023-07-23 12:11:15 CDT]** Added an even better bcround() for even more assured precision math.
* **[2023-07-23 12:10:09 CDT]** Upgraded the unit tests to PHP 8.
* **[2023-07-23 12:08:15 CDT]** [m] Updated the copyright notices.
* **[2023-07-23 12:05:45 CDT]** Upgraded to PHPUnit v10.
* **[2023-07-23 07:06:09 CDT]** Migrated to the phpexperts/money package name.

## v2.0.0: 2019-04-13

* **[2019-04-13]** [Major] Implemented strict_types so that precision is guaranteed. (#11)
* **[2019-04-13]** Added NumberHelper::assertIsNumeric() to enforce number-like values. (#10)
* **[2019-04-13]** Added arbitrary precision for converting decimals to lossless integers. (#9)
* **[2019-04-13]** [Major] Internalized the calculation strategy classes. (#7)
* **[2019-04-13]** [minor] Made pratically every class `final` to protected against bad devs. (#6)
* **[2019-04-13]** [minor] Upgraded to PHPUnit 8.0. (#8)

## v1.6.1: 2019-04-12

* **[2019-04-12]** Added a NumberHelper::isFloatLike() check utility. (#4)

## v1.6.0: 2019-04-10

* [Major] Added ::getWithFullPrecision() for seeing the whole number. 

## v1.5.0: A whole lot of goodness.  2019-04-10 

* [Major] Definitely require PHP 7.2, and BCMath for dev.
* [Major] (#1) Completely reimplemented bcround() and fixed serious bugs.
* [Major] Fixed the precision problems in BCMathCalcStrategy.
* [Major] Fixed the precision problems in NativeCalcStrategy.
* [Major] Added a whole bunch of type saftey to modding native numbers.
* [Major] Added 100% Unit Test code coverage. Hurray!
* [Major] Added the I_AM_A_DUMMY functionality to NativeCalc->modulus().
* [minor] Fixed the copyright headers.
* [minor] Added a mechanism to dynamically test whether extension paths work.
* [minor] Added functionality to find out what strategy Money is using.
* Reimplemented NativeCalcStrat's compare().
* Added phpstan and fixed every type discrepancy.
* Added TravisCI support.


tests/MoneyTest.php
<?php declare(strict_types=1);

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2019-2023 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  https://www.phpexperts.pro/
 *  https://github.com/PHPExpertsInc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType\Tests;

use PHPExperts\MoneyType\Internal\NumberHelper;
use PHPExperts\MoneyType\Money;
use PHPExperts\MoneyType\MoneyCalculationStrategy;
use PHPUnit\Framework\TestCase;

/** @testdox PHPExperts\MoneyType\Money */
final class MoneyTest extends TestCase
{
    private static function assertCalcStrategy(Money $moneyType, string $expected)
    {
        $actual = $moneyType->getStrategy();

        self::assertSame($expected, $actual);
    }

    public static function getNonNumericDataTypes(): array
    {
        return [
            'Non-numerical string' => 'asdf',
            'Semi-numeric string'  => '1 asdf',
            'Array'                => [],
            'Object'               => new \stdClass(),
            'Boolean'              => true,
            'Null'                 => null,
            'Resource'             => fopen('php://memory', 'r'),
            'Closure / Callable'   => function() {},
        ];
    }

    public static function buildMockCalcStrategy(): MoneyCalculationStrategy
    {
        return new class('asdf') implements MoneyCalculationStrategy
        {
            public function __toString(): string
            {
                return '0';
            }

            public function getWithFullPrecision(): string
            {
                return '1';
            }

            public function add(string $rightOperand): string
            {
                return '2';
            }

            public function subtract(string $rightOperand): string
            {
                return '3';
            }

            public function multiply(string $rightOperand): string
            {
                return '4';
            }

            public function divide(string $rightOperand): string
            {
                return '5';
            }

            public function modulus(string $rightOperand): string
            {
                return '6';
            }

            public function compare(string $rightOperand): int
            {
                return 7;
            }

            public function __construct(string $leftOperand)
            {
            }
        };
    }

    public function testCanOnlyBeInstantiatedWithANumericString()
    {
        self::assertInstanceOf(Money::class, new Money('5', self::buildMockCalcStrategy()));
    }

    public function testWillReportWhatStrategyIsBeingUsed()
    {
        $moneyType = new Money('5');
        self::assertCalcStrategy($moneyType, 'BCMathCalcStrategy');
    }

    /** @testdox Will use BCMath if it is available */
    public function testWillUseBCMathIfItIsAvailable()
    {
        $hasBCMath = function(): bool {
            return true;
        };

        $moneyType = new Money('5', null, $hasBCMath);
        self::assertCalcStrategy($moneyType, 'BCMathCalcStrategy');
    }

    public function testWillFallBackToNativePhpIfNecessary()
    {
        $hasBCMath = function() {
            return false;
        };

        $moneyType = new Money('5', null, $hasBCMath);
        self::assertCalcStrategy($moneyType, 'NativeCalcStrategy');
    }

    public function testProxiesEverythingToItsCalculationStrategy()
    {
        $mockCalcStrat = self::buildMockCalcStrategy();
        $moneyType = new Money('5', $mockCalcStrat);

        $calcStratOps = get_class_methods($mockCalcStrat);
        // Filter out the __construct method.
        $calcStratOps = array_diff($calcStratOps, ['__construct']);

        foreach ($calcStratOps as $index => $op) {
            self::assertEquals($index, $moneyType->$op('0'), $op);
        }
    }

    public function testConfirmThatTheReadmeDemoWorks()
    {
        $money = new Money('5.22');

        $money->add('0.55');
        self::assertSame('5.77', $money.'');

        # It keeps precision much much better than mere cents.
        $money->subtract('0.0001');
        self::assertSame('5.77', $money.'');

        $money->subtract('0.004');
        self::assertSame('5.77', $money.'');

        $money->subtract('0.001');
        self::assertSame('5.76', $money.'');

        $money->multiply('55.777355');
        self::assertSame('321.55', $money.'');
        self::assertSame('321.5508738395', $money->getWithFullPrecision());

        $money->divide('1.000005');
        self::assertSame('321.55', $money.'');
        self::assertSame('321.5492660931', $money->getWithFullPrecision());

        self::assertSame(0,  $money->compare('321.5492660931')); //  0 = equal
        self::assertSame(-1, $money->compare('321.5492660930')); // -1 = less
        self::assertSame(1,  $money->compare('321.5492660932')); //  1 = more

        $btc = '1.55527331';
        $sats = 155527331;
        self::assertSame($sats, NumberHelper::convertToCents($btc, 8));
    }
}



tests/BCMathCalcStrategyTest.php
<?php

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2019 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  https://www.phpexperts.pro/
 *  https://github.com/phpexpertsinc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType\Tests;

use PHPExperts\MoneyType\BCMathCalcStrategy;
use function PHPExperts\MoneyType\bcround;
use PHPUnit\Framework\TestCase;

final class BCMathCalcStrategyTest extends TestCase
{
    use MathCalcStrategyTestBase;

    public function setUp()
    {
        parent::setUp();

        $this->calcStratName = BCMathCalcStrategy::class;
        $this->calcStrat = new BCMathCalcStrategy('1.12');
    }

    public function testCanGetTheFullPrecisionValueToMoreThanSixteenDecimals()
    {
        $money = new BCMathCalcStrategy('56.14111331312555551');
        self::assertSame('56.14111331312555551', $money->getWithFullPrecision());
    }

    public function testCanAddWithHighPrecision()
    {
        $this->calcStrat->add('55.021113313');
        self::assertSame('56.1411133131', $this->calcStrat->add('0.0000000001'));
    }

    public function testCanSubtractWithHighPrecision()
    {
        self::assertSame('-53.9011133130', $this->calcStrat->subtract('55.021113313'));
    }

    public function testCanMultiplyWithHighPrecision()
    {
        self::assertSame('61.6236469105', $this->calcStrat->multiply('55.021113313'));
    }

    public function testCanDivideWithHighPrecision()
    {
        self::assertSame('53.0470987665', $this->calcStrat->divide('0.021113313'));
    }

    public function testCanCompareTwoNumbersWithHighPrecision()
    {
        // Assert same with extra precision.
        self::assertSame(0, $this->calcStrat->compare('1.120000000'));

        // Assert is less with extra precision.
        self::assertSame(-1, $this->calcStrat->compare('1.119999999'));

        // Assert is more with extra precision.
        self::assertSame(1, $this->calcStrat->compare('1.120000001'));
    }

    public function testCanComputeHighPrecisionModulus()
    {
        self::assertSame('1.1200000000', $this->calcStrat->modulus('2'));
    }

    public function testCanComputeTheModulusOfDecimals()
    {
        self::assertSame('0.0699993000', $this->calcStrat->modulus('0.1500001'));
    }

    public function testCanRoundWithHighPrecision()
    {
        // Test high precision.
        $expected = '-54.03';
        $actual = bcround('-54.0300000000', 2);
        self::assertSame($expected, $actual);

        // Test no precision.
        $expected = '-54.00';
        $actual = bcround('-54', 2);
        self::assertSame($expected, $actual);
    }
}


tests/NativeCalcStrategyTest.php
<?php

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2019 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  https://www.phpexperts.pro/
 *  https://github.com/phpexpertsinc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType\Tests;

use InvalidArgumentException;
use PHPExperts\MoneyType\NativeCalcStrategy;
use PHPUnit\Framework\TestCase;

final class NativeCalcStrategyTest extends TestCase
{
    use MathCalcStrategyTestBase;

    /** @var NativeCalcStrategy */
    protected $calcStrat;

    public function setUp()
    {
        parent::setUp();

        $this->calcStratName = NativeCalcStrategy::class;
        $this->calcStrat = new NativeCalcStrategy('1.12');
    }

    public function testWontAttemptOperationsWithNonNumbers()
    {
        // @FIXME: This functionality needs to be expanded to EVERY operation -and- strategy.
        try {
            $this->calcStrat->modulus('Not A Number');
            $this->fail('It somehow tried to compute a non-number modulus.');
        }
        catch (InvalidArgumentException $e) {
            $this->assertEquals("This is not a valid number.", $e->getMessage());
        }
    }

    public function testCanGetTheFullPrecisionValueToTwoDecimals()
    {
        $money = new NativeCalcStrategy('56.14111331312555551');
        self::assertSame('56.14', $money->getWithFullPrecision());
    }

    public function testCannotComputeTheModulusOfDecimals()
    {
        try {
            $this->calcStrat->modulus('0.1500001');
            $this->fail('It somehow tried to compute a decimal modulus.');
        }
        catch (InvalidArgumentException $e) {
            self::assertEquals('Cannot compute non-integer moduli: Install ext-bcmath.', $e->getMessage());
        }
    }

    public function testWillThrowAnExceptionIfAskedToComputeAnIntegerModulus()
    {
        $this->expectException('BadMethodCallException');
        $this->calcStrat->modulus('2');
    }

    public function testCanComputeTheModulusOfIntegersIfDevPassesIAmADummyParameter()
    {
        self::assertSame('1', $this->calcStrat->modulus('2', NativeCalcStrategy::I_AM_A_DUMMY));
    }
}


src/Money.php
<?php declare(strict_types=1);

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2017-2023 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  @ 2017-05-01 14:17 IST
 *  https://www.phpexperts.pro/
 *  https://github.com/PHPExpertsInc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType;

use PHPExperts\MoneyType\Internal\BCMathCalcStrategy;
use PHPExperts\MoneyType\Internal\NativeCalcStrategy;
use ReflectionClass;
use ReflectionException;

final class Money implements MoneyCalculationStrategy
{
    protected $strategy;

    public function __construct(string $amount, MoneyCalculationStrategy $calcStrategy = null, callable $hasBCMath = null)
    {
        if (!$hasBCMath) {
            $hasBCMath = function(): bool {
                return extension_loaded('bcmath');
            };
        }

        if (!$calcStrategy) {
            if ($hasBCMath()) {
                $calcStrategy = new BCMathCalcStrategy($amount);
            } else {
                $calcStrategy = new NativeCalcStrategy($amount);
            }
        }
        $this->strategy = $calcStrategy;
    }

    public function getWithFullPrecision(): string
    {
        return $this->strategy->getWithFullPrecision();
    }

    /**
     * Outputs the dollars and cents of the object.
     *
     * @return string
     */
    public function __toString(): string
    {
        return $this->strategy->__toString();
    }

    /**
     * Returns the strategy used.
     *
     * @return string
     * @throws ReflectionException
     */
    public function getStrategy(): string
    {
        return (new ReflectionClass($this->strategy))->getShortName();
    }

    public function add($rightOperand): string
    {
        return $this->strategy->add($rightOperand);
    }

    public function subtract($rightOperand): string
    {
        return $this->strategy->subtract($rightOperand);
    }

    public function multiply($rightOperand): string
    {
        return $this->strategy->multiply($rightOperand);
    }

    public function divide($rightOperand): string
    {
        return $this->strategy->divide($rightOperand);
    }

    public function modulus($rightOperand): string
    {
        return $this->strategy->modulus($rightOperand);
    }

    public function compare($rightOperand): int
    {
        return $this->strategy->compare($rightOperand);
    }
}


src/Internal/NumberHelper.php
<?php declare(strict_types=1);

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2017-2023 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  @ 2017-05-01 14:38 IST
 *  https://www.phpexperts.pro/
 *  https://github.com/PHPExpertsInc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType\Internal;

use InvalidArgumentException;

/**
 * A collection of static functions for number manipulation.
 */
class NumberHelper
{
    /**
     * Asserts if the input is really a numeric string or throws an exception.
     *
     * @param  float|int|string $number The supposedly numeric value.
     * @throws InvalidArgumentException when the input is not a numeric string.
     */
    public static function assertIsNumeric($number): void
    {
        if (!is_numeric($number)) {
            throw new InvalidArgumentException('This is not a parsable numeric string.');
        }
    }

    /**
     * Determines if a variable appears to be a float or not.
     *
     * @param string|int|double $number
     * @return bool True if it appears to be an integer value. "75.0000" returns false.
     * @throws InvalidArgumentException if $number is not a valid number.
     */
    public static function isFloatLike($number): bool
    {
        // Bail if it isn't even a number.
        self::assertIsNumeric($number);

        // Try to convert the variable to a float.
        $floatVal = floatval($number);

        // If the parsing succeeded and the value is not equivalent to an int, it's probably a float.
        return ($floatVal && intval($floatVal) != $floatVal);
    }

    /**
     * Converts a float-like string to an int with no precision loss.
     *
     * @param  string $float
     * @param  int    $precision
     * @return int
     */
    public static function convertToCents(string $float, int $precision = 2): int
    {
        return (int) round((float)$float * 10**$precision);
    }
}


src/MoneyCalculationStrategy.php
<?php declare(strict_types=1);

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2017-2023 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  @ 2017-05-01 14:31 IST
 *  https://www.phpexperts.pro/
 *  https://github.com/PHPExpertsInc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType;

interface MoneyCalculationStrategy
{
    public function __construct(string $leftOperand);
    public function __toString(): string;
    public function getWithFullPrecision(): string;
    public function add(string $rightOperand): string;
    public function subtract(string $rightOperand): string;
    public function multiply(string $rightOperand): string;
    public function divide(string $rightOperand): string;
    public function modulus(string $rightOperand): string;
    public function compare(string $rightOperand): int;
}


src/Internal/NativeCalcStrategy.php
<?php declare(strict_types=1);

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2017-2023 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  @ 2017-05-01 14:38 IST
 *  https://www.phpexperts.pro/
 *  https://github.com/PHPExpertsInc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType\Internal;

use BadMethodCallException;
use InvalidArgumentException;
use PHPExperts\MoneyType\MoneyCalculationStrategy;

/**
 * @internal
 */
final class NativeCalcStrategy implements MoneyCalculationStrategy
{
    public const I_AM_A_DUMMY = 'But I really need this';

    private $leftOperand;

    public function __construct(string $leftOperand)
    {
        NumberHelper::assertIsNumeric($leftOperand);

        $this->leftOperand = NumberHelper::convertToCents($leftOperand);
    }

    public function getWithFullPrecision(): string
    {
        return (string)($this->leftOperand / 100);
    }

    public function __toString(): string
    {
        return (string)round($this->leftOperand / 100, 2);
    }

    public function add(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $rightOperand = NumberHelper::convertToCents($rightOperand);
        $this->leftOperand += $rightOperand;

        return (string)round($this->leftOperand / 100, 2);
    }

    public function subtract(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $rightOperand = NumberHelper::convertToCents($rightOperand);
        $this->leftOperand -= $rightOperand;
        return (string)round($this->leftOperand / 100, 2);
    }

    public function multiply(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $rightOperand = NumberHelper::convertToCents($rightOperand);
        $this->leftOperand *= $rightOperand / 100;

        return (string)round($this->leftOperand / 100, 2);
    }

    public function divide(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $rightOperand = NumberHelper::convertToCents($rightOperand);
        $this->leftOperand /= $rightOperand / 100;

        return (string)round($this->leftOperand / 100, 2);
    }

    /**
     * @param string $rightOperand It must be a float-like
     * @param string $iAmStupid Do you really want an imprecise modulus with a precision package??
     * @return string
     */
    public function modulus(string $rightOperand, string $iAmStupid = ''): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        if (NumberHelper::isFloatLike($rightOperand)) {
            throw new InvalidArgumentException('Cannot compute non-integer moduli: Install ext-bcmath.');
        }

        if ($iAmStupid !== self::I_AM_A_DUMMY) {
            throw new BadMethodCallException(
                'You CANNOT get precise a modulus answer via Native PHP. ' . "\n" .
                'Install ext-bcmath (preferred) or set $iAmStupid to true.'
            );
        }

        return (string) (($this->leftOperand / 100) % $rightOperand);
    }

    /**
     * @param string $rightOperand
     * @return int 0 if equal, -1 if $rightOperand is less, 1 if it is greater.
     */
    public function compare(string $rightOperand): int
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $rightOperand = NumberHelper::convertToCents($rightOperand);

        return $rightOperand <=> $this->leftOperand;
    }
}


src/Internal/BCMathCalcStrategy.php
<?php declare(strict_types=1);

/**
 * This file is part of MoneyType, a PHP Experts, Inc., Project.
 *
 * Copyright © 2017-2023 PHP Experts, Inc.
 * Author: Theodore R. Smith <theodore@phpexperts.pro>
 *  GPG Fingerprint: 4BF8 2613 1C34 87AC D28F  2AD8 EB24 A91D D612 5690
 *  @ 2017-05-01 14:38 IST
 *  https://www.phpexperts.pro/
 *  https://github.com/PHPExpertsInc/MoneyType
 *
 * This file is licensed under the MIT License.
 */

namespace PHPExperts\MoneyType\Internal;

use PHPExperts\MoneyType\MoneyCalculationStrategy;

/**
 * @internal
 */
final class BCMathCalcStrategy implements MoneyCalculationStrategy
{
    public const PRECISION = 10;

    private $leftOperand;

    public function __construct(string $leftOperand)
    {
        NumberHelper::assertIsNumeric($leftOperand);

        $this->leftOperand = $leftOperand;
    }

    public function getWithFullPrecision(): string
    {
        return $this->leftOperand;
    }

    public function __toString(): string
    {
        return bcround($this->leftOperand, 2);
    }

    public function add(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $this->leftOperand = bcadd($this->leftOperand, $rightOperand, self::PRECISION);

        return $this->leftOperand;
    }

    public function subtract(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $this->leftOperand = bcsub($this->leftOperand, $rightOperand, self::PRECISION);

        return $this->leftOperand;
    }

    public function multiply(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $this->leftOperand = bcmul($this->leftOperand, $rightOperand, self::PRECISION);

        return $this->leftOperand;
    }

    public function divide(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        $this->leftOperand = bcdiv($this->leftOperand, $rightOperand, self::PRECISION);

        return $this->leftOperand;
    }

    public function modulus(string $rightOperand): string
    {
        NumberHelper::assertIsNumeric($rightOperand);

        return bcmod($this->leftOperand, $rightOperand, self::PRECISION);
    }

    /**
     * @param string $rightOperand
     * @return int 0 if equal, -1 if $rightOperand is less, 1 if it is greater.
     */
    public function compare(string $rightOperand): int
    {
        NumberHelper::assertIsNumeric($rightOperand);

        return bccomp($rightOperand, $this->leftOperand, self::PRECISION);
    }
}

/**
 * Based off of https://stackoverflow.com/a/1653826/430062
 * Thanks, [Alix Axel](https://stackoverflow.com/users/89771/alix-axel)!
 *
 * @param string $number
 * @param int $precision
 * @return string
 */
function bcround(string $n, $p = BCMathCalcStrategy::PRECISION)
{
    $e = bcpow('10', (string) ($p + 1));
    $result = bcdiv(bcadd(bcmul($n, $e, 0), strpos($n, '-') === 0 ? '-5' : '5'), $e, $p);

    // Pad it out to the desired precision.
    return bcadd($result, '0', $p);
}

function bcround_v1(string $number, int $precision = BCMathCalcStrategy::PRECISION): string
{
    NumberHelper::assertIsNumeric($number);

    // Pad it out to the desired precision.
    if ($number[0] != '-') return bcadd($number, '0.' . str_repeat('0', $precision) . '5', $precision);
    return bcsub($number, '0.' . str_repeat('0', $precision) . '5', $precision);
}
